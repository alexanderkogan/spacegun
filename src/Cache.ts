export class Cache<S, T> {
    private data: Map<S, { value: T, age: number }> = new Map()

    public constructor(
        public maxAgeSeconds: number,
        public maxSize: number = 100
    ) { }

    public get(key: S): T | undefined {
        const entry = this.data.get(key)
        if (entry !== undefined && Date.now() - entry.age < this.maxAgeSeconds * 1000) {
            return entry.value
        }
        return undefined
    }

    public set(key: S, value: T): void {
        this.data.set(key, { value, age: Date.now() })
    }

    public async calculate(key: S, valueProvider: () => Promise<T>): Promise<T> {
        let value = this.get(key)
        if (value === undefined) {
            value = await valueProvider()
            this.set(key, value)
        }
        return value
    }
}
