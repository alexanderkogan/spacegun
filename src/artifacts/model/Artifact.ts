export interface Artifact {
    name: string
    data: object
}
