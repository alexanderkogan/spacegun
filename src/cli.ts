import { parse } from "./options"
import { KubernetesClusterProvider } from './cluster/kubernetes/KubernetesCluster'
import { commands } from "./commands"

(async () => {
    const options = parse()
    const cluster = new KubernetesClusterProvider(options.kube)
    try {
        await commands[options.command](cluster)
    } catch (error) {
        console.error(`Command ${options.command} failed: ${error}`)
    }
})()
