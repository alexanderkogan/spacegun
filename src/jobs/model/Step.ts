import { Filter } from "./Filter"

export type StepType = "clusterProbe" | "planImageDeployment" | "planClusterDeployment" | "applyDeployment" | "takeSnapshot" | "rollback" | "logError"

export interface StepDescription {
    readonly name: string
    readonly type: StepType
    readonly onSuccess?: string
    readonly onFailure?: string

    readonly filter?: Partial<Filter>
    readonly tag?: string       // planImageDeployment
    readonly semanticTagExtractor?: string // planImageDeployment
    readonly cluster?: string   // planClusterDeployment
    readonly hook?: string      // clusterProbe
}
