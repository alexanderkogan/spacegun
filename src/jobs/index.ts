import { safeLoad } from "js-yaml"
import { readFileSync, readdirSync } from "fs"

import { Job } from "@/jobs/model/Job"
import { JobSource } from "@/jobs/model/JobSource"

export function load(path: string = "./jobs"): Map<string, Job> {
    const files = readdirSync(path)
    const jobs: Map<string, Job> = new Map()
    for (const file of files) {
        const name = file.split(".")[0]
        const filePath = `${path}/${file}`
        const fileContent = readFileSync(filePath, 'utf8')
        const partial = safeLoad(fileContent) as Partial<Job>
        jobs.set(name, validateJob(partial, name))
    }
    return jobs
}

export function validateJob(partial: Partial<Job>, name: string): Job {
    if (partial.cluster === undefined) {
        throw new Error(`job ${name} must contain a cluster`)
    }
    if (partial.from === undefined) {
        throw new Error(`job ${name} must contain a source`)
    }

    const from = validateJobSource(partial.from, name, partial.cluster)
    return { name, cluster: partial.cluster, from, cron: partial.cron }
}

export function validateJobSource(partial: Partial<JobSource>, name: string, cluster: string): JobSource {
    let source: JobSource
    switch (partial.type) {
        case "cluster": {
            if (partial.expression === undefined) {
                throw new Error(`the source of job ${name} must have a cluster expression`)
            }
            if (partial.expression === cluster) {
                throw new Error(`the source of job ${name} cannot be its target`)
            }
            source = { type: "cluster", expression: partial.expression }
            break
        }
        case "image": {
            if (partial.expression === undefined) {
                source = { type: "image", expression: "^.*$" }
            } else {
                source = { type: "image", expression: partial.expression }
            }
            break
        }
        default:
            throw new Error(`${partial.type} of job ${name} is not a valid source type`)
    }
    return source
}
