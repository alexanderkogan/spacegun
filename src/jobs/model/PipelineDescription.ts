import { StepDescription } from './Step'

export interface PipelineDescription {
    readonly name: string
    readonly cluster: string
    readonly cron?: string
    readonly start: string
    readonly steps: StepDescription[]
}
