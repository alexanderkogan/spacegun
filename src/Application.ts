import { load as loadConfig, Config } from "@/config"
import { commands, printHelp } from "@/commands"
import { parse } from "@/options"

import { Layers } from "@/dispatcher/model/Layers"
import { run as runDispatcher } from "@/dispatcher"

import { IO } from "@/IO"
import { CronRegistry } from "@/crons/CronRegistry"

import { GitConfigRepository, fromConfig as gitRepoFromConfig } from "@/config/git/GitConfigRepository"
import { fromConfig as configRepoFromConfig } from "@/config/filesystem/FilesystemConfigRepository";

import { KubernetesClusterRepository } from "@/cluster/kubernetes/KubernetesClusterRepository"
import { DockerImageRepository } from "@/images/docker/DockerImageRepository"
import { JobsRepositoryImpl } from "@/jobs/JobsRepositoryImpl"
import { SlackEventRepository } from "@/events/slack/SlackEventRepository"

import { init as initCluster } from "@/cluster/ClusterModule"
import { init as initEvents } from "@/events/EventModule"
import { init as initImages } from "@/images/ImageModule"
import { init as initConfig } from "@/config/ConfigModule"
import { init as initJobs } from "@/jobs/JobsModule"
import { init as initViews } from "@/views"

import { Options } from "@/options"

export class Application {

    public static create(): Application {
        const io = new IO()
        const crons = new CronRegistry()
        const options = parse()
        return new Application(io, crons, options)
    }

    public constructor(
        public readonly io: IO,
        public readonly crons: CronRegistry,
        public readonly options: Options
    ) { }

    public async run() {
        try {
            const config = loadConfig(this.options.config)
            this.initialize(config)

            runDispatcher(config.server.host, config.server.port)
            if (process.env.LAYER === Layers.Standalone || process.env.LAYER === Layers.Client) {
                await commands[this.options.command](this.io)
            }
        } catch (e) {
            printHelp(this.io, e)
        }
    }

    public async checkForConfigChange(git: GitConfigRepository) {
        const hasNewConfig = await git.hasNewConfig()
        if (hasNewConfig) {
            await git.fetchNewConfig()
            this.crons.removeAllCrons()
            this.reload()
        }
    }

    private reload(): void {
        try {
            const config = loadConfig(this.options.config)
            this.initialize(config)
        } catch (e) {
            this.io.out(`could not reload config ${e.message}`)
        }
    }

    private initialize(config: Config) {
        const gitRepo = gitRepoFromConfig(config)
        if (gitRepo !== undefined) {
            this.crons.register(
                "config-reload",
                config.git!.cron,
                () => this.checkForConfigChange(gitRepo)
            )
            initConfig(gitRepo)
        } else {
            const fileSystemRepo = configRepoFromConfig(config)
            initConfig(fileSystemRepo)
        }

        initViews(config)
        initEvents([
            SlackEventRepository.fromConfig(config.slack)
        ])
        initCluster(KubernetesClusterRepository.fromConfig(config.kube, config.namespaces))
        initImages(DockerImageRepository.fromConfig(config.docker))
        initJobs(JobsRepositoryImpl.fromConfig(config.jobs, this.crons))
    }
}
