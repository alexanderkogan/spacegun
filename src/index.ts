import { Layers } from "@/dispatcher/model/Layers"

import { load as loadConfig } from "@/config"
import { parse } from "@/options"
import { run as runDispatcher } from "@/dispatcher"
import { commands, printHelp } from "@/commands"
import { IO } from "@/IO"

import { init as initJobs } from "@/jobs/JobsModule"
import { init as initCluster } from "@/cluster/ClusterModule"
import { init as initImages } from "@/images/ImageModule"

import { KubernetesClusterRepository } from "@/cluster/kubernetes/KubernetesClusterRepository"
import { DockerImageRepository } from "@/images/docker/DockerImageRepository"
import { JobsRepositoryImpl } from "@/jobs/JobsRepositoryImpl"

const options = parse()
const config = loadConfig(options.config)
const io = new IO()
try {
    if (config instanceof Error) {
        printHelp(io, true)
    } else {
        initJobs(JobsRepositoryImpl.fromConfig(config.jobs))
        initCluster(KubernetesClusterRepository.fromConfig(config.kube))
        initImages(DockerImageRepository.fromConfig(config.docker))

        runDispatcher(config.server.host, config.server.port)
        if (process.env.LAYER === Layers.Standalone || process.env.LAYER === Layers.Client) {
            commands[options.command](io)
        }
    }
} catch (e) {
    console.error(e)
}
