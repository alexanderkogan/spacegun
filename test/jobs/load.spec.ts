import { load, validatePipeline, validateSteps } from '../../src/jobs'
import { PipelineDescription } from '../../src/jobs/model/PipelineDescription'
import { StepDescription } from '../../src/jobs/model/Step'

describe('job loading', () => {
    const pipelines: Map<string, PipelineDescription> = load(`${__dirname}/pipelines`)

    it('loads jobs', () => {
        expect(pipelines.get('pipeline1')).toEqual({
            name: 'pipeline1',
            cluster: 'cluster1',
            cron: '0 */5 * * * MON-FRI',
            start: 'probe1',
            steps: [
                {
                    name: 'probe1',
                    type: 'clusterProbe',
                    hook: 'someHook',
                },
                {
                    name: 'plan1',
                    type: 'planImageDeployment',
                    tag: 'latest',
                    onSuccess: 'apply1',
                    filter: { deployments: ['deployment1', 'deployment2'], namespaces: ['namespace1', 'namespace2'] },
                },
                {
                    name: 'plan2',
                    type: 'planImageDeployment',
                    semanticTagExtractor: '/^\\d{4}\\-\\d{1,2}\\-\\d{1,2}$',
                    onSuccess: 'apply1',
                },
                { name: 'apply1', type: 'applyDeployment' },
            ],
        })
        expect(pipelines.get('pipeline2')).toEqual({
            cluster: 'cluster2',
            cron: '0 */5 * * * MON-FRI',
            name: 'pipeline2',
            start: 'probe1',
            steps: [
                { name: 'probe1', type: 'clusterProbe', hook: 'https://some.hook.com', onSuccess: 'deployImage' },
                {
                    name: 'plan1',
                    type: 'planClusterDeployment',
                    cluster: 'cluster3',
                    onFailure: 'rollback1',
                    onSuccess: 'apply1',
                },
                { name: 'apply1', type: 'applyDeployment', onFailure: 'rollback1', onSuccess: 'snapshot1' },
                { name: 'snapshot1', type: 'takeSnapshot' },
                { name: 'rollback1', type: 'rollback' },
            ],
        })
        expect(pipelines.get('namespacePipeline')).toEqual({
            cluster: 'targetCluster',
            name: 'namespacePipeline',
            start: 'plan',
            steps: [
                {
                    name: 'plan',
                    type: 'planNamespaceDeployment',
                    cluster: 'sourceCluster',
                    source: 'namespaceSource',
                    target: 'namespaceTarget',
                    filter: {
                        deployments: ['deployment1', 'deployment2'],
                    },
                    onFailure: 'error',
                    onSuccess: 'apply',
                },
                { name: 'apply', type: 'applyDeployment' },
                { name: 'error', type: 'logError' },
            ],
        })
    })
})

describe('validatePipeline', () => {
    it('ensures cluster exists', () => {
        expect(() => validatePipeline({}, 'jobName')).toThrowErrorMatchingSnapshot()
    })

    it('ensures steps exists', () => {
        expect(() => validatePipeline({ cluster: 'someCluster' }, 'jobName')).toThrowErrorMatchingSnapshot()
    })

    it('ensures that a start step exists', () => {
        expect(() => validatePipeline({ cluster: 'someCluster', steps: [] }, 'jobName')).toThrowErrorMatchingSnapshot()
    })

    it('ensures that a start step exists', () => {
        expect(validatePipeline({ cluster: 'someCluster', steps: [], start: 'someStart' }, 'jobName')).toEqual({
            cluster: 'someCluster',
            name: 'jobName',
            start: 'someStart',
            steps: [],
        })
    })
})

describe('validateSteps', () => {
    it('validates empty array', () => {
        expect(validateSteps([], 'jobname', 'clustername')).toEqual([])
    })

    it('ensures steps have name', () => {
        expect(() => validateSteps([{}], 'jobname', 'clustername')).toThrowErrorMatchingSnapshot()
    })

    it('ensures steps have a type', () => {
        expect(() => validateSteps([{ name: 'stepName' }], 'jobname', 'clustername')).toThrowErrorMatchingSnapshot()
    })

    describe('planClusterDeployment', () => {
        it('ensures cluster exists', () => {
            expect(() =>
                validateSteps([{ name: 'stepName', type: 'planClusterDeployment' }], 'jobname', 'clustername')
            ).toThrowErrorMatchingSnapshot()
        })

        it('ensures cluster is not origin', () => {
            expect(() =>
                validateSteps(
                    [
                        {
                            name: 'stepName',
                            type: 'planClusterDeployment',
                            cluster: 'clustername',
                        },
                    ],
                    'jobname',
                    'clustername'
                )
            ).toThrowErrorMatchingSnapshot()
        })

        it('validates step', () => {
            const step: StepDescription = {
                name: 'stepName',
                type: 'planClusterDeployment',
                cluster: 'otherCluster',
                onFailure: 'failure',
                onSuccess: 'success',
            }
            expect(validateSteps([step], 'jobname', 'clustername')).toEqual([step])
        })
    })

    describe('planNamespaceDeployment', () => {
        it('ensures filter.namespaces is not set', () => {
            expect(() =>
                validateSteps(
                    [
                        {
                            name: 'stepName',
                            type: 'planNamespaceDeployment',
                            source: 'sourceNamespace',
                            target: 'targetNamespace',
                            filter: {
                                namespaces: ['sourceNamespace'],
                            },
                        },
                    ],
                    'jobname',
                    'clustername'
                )
            ).toThrowErrorMatchingSnapshot()
        })

        it('ensures source namespace is set', () => {
            expect(() =>
                validateSteps(
                    [
                        {
                            name: 'stepName',
                            type: 'planNamespaceDeployment',
                            target: 'targetNamespace',
                        },
                    ],
                    'jobname',
                    'clustername'
                )
            ).toThrowErrorMatchingSnapshot()
        })

        it('ensures target namespace is set', () => {
            expect(() =>
                validateSteps(
                    [
                        {
                            name: 'stepName',
                            type: 'planNamespaceDeployment',
                            source: 'sourceNamespace',
                        },
                    ],
                    'jobname',
                    'clustername'
                )
            ).toThrowErrorMatchingSnapshot()
        })

        it('validates step', () => {
            const step: StepDescription = {
                name: 'stepName',
                type: 'planNamespaceDeployment',
                cluster: 'cluster',
                source: 'source',
                target: 'target',
                filter: {
                    resources: ['deployment1', 'deployment2'],
                },
                onFailure: 'failure',
                onSuccess: 'success',
            }
            expect(validateSteps([step], 'jobname', 'clustername')).toEqual([step])
        })

        it('uses target cluster as source cluster if no source cluster given', () => {
            const input: StepDescription = {
                name: 'stepName',
                type: 'planNamespaceDeployment',
                source: 'source',
                target: 'target',
                filter: {
                    resources: ['deployment1', 'deployment2'],
                },
                onFailure: 'failure',
                onSuccess: 'success',
            }
            const expected: StepDescription = { ...input, cluster: 'clustername' }
            expect(validateSteps([input], 'jobname', 'clustername')).toEqual([expected])
        })
    })

    describe('clusterProbe', () => {
        it('ensures hook exists', () => {
            expect(() =>
                validateSteps([{ name: 'stepName', type: 'clusterProbe' }], 'jobname', 'clustername')
            ).toThrowErrorMatchingSnapshot()
        })

        it('validates step', () => {
            const step: StepDescription = {
                name: 'stepName',
                type: 'clusterProbe',
                hook: 'hook',
                onFailure: 'failure',
                onSuccess: 'success',
            }
            expect(validateSteps([step], 'jobname', 'clustername')).toEqual([step])
        })
    })

    describe('planImageDeployment', () => {
        it('validates step', () => {
            const step: StepDescription = {
                name: 'stepName',
                type: 'planImageDeployment',
                onFailure: 'failure',
                onSuccess: 'success',
            }
            expect(validateSteps([step], 'jobname', 'clustername')).toEqual([step])
        })
    })
})
