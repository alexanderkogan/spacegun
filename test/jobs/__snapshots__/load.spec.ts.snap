// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validatePipeline ensures cluster exists 1`] = `"job jobName must contain a cluster"`;

exports[`validatePipeline ensures steps exists 1`] = `"job jobName must contain steps"`;

exports[`validatePipeline ensures that a start step exists 1`] = `"job jobName must contain an start step"`;

exports[`validateSteps clusterProbe ensures hook exists 1`] = `"in step stepName of job jobname the probe is missing"`;

exports[`validateSteps ensures steps have a type 1`] = `"undefined of job jobname is not a valid step type"`;

exports[`validateSteps ensures steps have name 1`] = `"a step of job jobname has no name"`;

exports[`validateSteps planClusterDeployment ensures cluster exists 1`] = `"in step stepName of job jobname the cluster expression is missing"`;

exports[`validateSteps planClusterDeployment ensures cluster is not origin 1`] = `"in step stepName of job jobname the cluster expression cannot be the source"`;

exports[`validateSteps planImageDeployment ensures hook exists 1`] = `"in step stepName of job jobname the tag is missing"`;
